XMonad/Hooks/DynamicLog.hs

```diff
@@ -433,32 +445,34 @@ statusBar' :: LayoutClass l Window
                            -- ^ The desired key binding to toggle bar visibility
            -> XConfig l    -- ^ The base config
            -> IO (XConfig (ModifiedLayout AvoidStruts l))
-statusBar' cmd xpp k conf = do
-    h <- spawnPipe cmd
-    makeStatusBar
-        (xpp >>= \pp -> dynamicLogWithPP pp{ ppOutput = hPutStrLn h })
-        mempty
-        k
-        conf
-
--- | Helper function to make status bars.  This should not be used on its own;
--- use 'statusBarProp' or 'statusBar' (or their respective variants) instead.
+statusBar' = makeSBWithKey .: statusBarHandleConfig'
+
+-- | Incorporates a 'StatusBarConfig' into an 'XConfig' by taking care of the
+-- necessary plumbing.
 makeStatusBar :: LayoutClass l Window
-              => X ()      -- ^ 'logHook' to execute
-              -> X ()      -- ^ 'startupHook' to execute
-              -> (XConfig Layout -> (KeyMask, KeySym))
-                           -- ^ The desired key binding to toggle bar visibility
-              -> XConfig l -- ^ The base config
+              =>  XConfig l       -- ^ The base config
+              -> StatusBarConfig -- ^ The status bar config
               -> IO (XConfig (ModifiedLayout AvoidStruts l))
```

Comment:
The order of parameters is a bit weird here. It seems to be optimized for the following usecase:
```
main = xmonad =<< makeStatusBar myConf =<< mySB
```
but it quickly gets ugly if one doesn't have a toplevel def for myConf, like in my config: https://github.com/liskin/dotfiles/blob/95495d2966543893349c37c61361a38c1fb104d4/.xmonad/xmonad.hs#L508

Reply:
It was actually different, the `StatusBarConfig` was the first parameter. @slotThe  and I discussed it a bit (https://github.com/xmonad/xmonad-contrib/pull/443#discussion_r556565592) 

The original `StatusBarConfig -> XConfig l -> IO (XConfig (ModifiedLayout AvoidStruts l))` resulted in constantly using `flip makeStatusBar conf` in the code, which was  bit weird.

Do you think we should add more documentation on how to use this with a config like yours? Or provide another more convenient function, just for a better UX? Both options seem valid to me

Reply:
Oh, I see. But I think the real problem wasn't the `flip`. The real problem was that you almost always had `IO StatusBarConfig` but `makeStatusBar` takes `StatusBarConfig`, so you had to wrestle with `>>=` or something. But this would've been fine, too:
```
do sb <- mySB
   xmonad =<< (makeStatusBar sb . ewmh . docks . … $ def{ … })
```
I mean, not totally fine, the IO in makeStatusBar is unnecessary, so it should really be:
```
do sb <- mySB
   xmonad $ makeStatusBar sb . ewmh . docks . … $ def{ … }
```

Or maybe keep the IO but use it for something: if all the functions that return StatusBarConfig return it in IO, then it's probably good UX to have makeStatusBar evaluate it. So then you have
```
xmonad =<< makeStatusBar mySB (ewmh . docks . def{ … })
```
You can do that even now with
```
xmonad =<< combineSBs [mySB] (ewmh . docks . def{ … })
```
but it discards the keybinding stuff so it's not a direct replacement. But that's another issue that I believe should be addressed. It's really be nice if these two were actually equivalent.

I haven't really thought this all through to the end, it needs some experimenting and maybe it's a dead end.

Reply:
I wanted `makeStatusBar` to return a pure XConfig, but I'm not sure how to do that. The main problem is with logging with pipes. You would need to have access to the pipe so you can write to it, so essentially, you'd need to spawn the status bar before creating the config. 

I think changing the parameter to an `IO StatusBarConf` would provide a better interface

Reply:
Yeah but you're not creating the pipes in `makeStatusBar` any more (or ever). `makeStatusBar` is literally `pure something`.

Reply:
Right, yes, silly me :sweat_smile:  I think it's between
```haskell
do sb <- mySB
   xmonad $ makeStatusBar sb . ewmh . docks . … $ def{ … }
``` 
or
```haskell
xmonad =<< makeStatusBar mySB (ewmh . docks . def{ … })
```
The second one is more convenient to write, but it maybe is uglier since the `IO` is not really needed. Both seem valid to me, maybe we say what you or others prefer